File: python_roguelike/data/weighted_choice.py
import random
from dataclasses import dataclass
from typing import TypeVar, Generic, List, Optional

T = TypeVar("T")


@dataclass
class WeightedChoice(Generic[T]):
    item: T
    weight: int

    def __post_init__(self):
        if self.weight <= 0:
            import sys
            print("weight must be > 0", file=sys.stderr)


def weighted_random_pick(choices: List[WeightedChoice], rng: random.Random):
    if not choices:
        return None
    total_weight = sum(c.weight for c in choices)
    if total_weight <= 0:
        return None
    rand_num = rng.randint(0, total_weight - 1)
    cumulative = 0
    for choice in choices:
        cumulative += choice.weight
        if rand_num < cumulative:
            return choice.item
    return choices[-1].item

File: src/Roguelike/Core/Data/WeightedChoice.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Roguelike.Data
{


    public class WeightedChoice<T>
    {
        public T Item { get; set; }
        public int Weight { get; set; }

        public WeightedChoice(T item, int weight)
        {
            if (weight <= 0)
            {
                Console.Error.WriteLine("weight must be > 0");
            }
            Item = item;
            Weight = weight;
        }
    }


    public static class WeightedRandom
    {


        public static T Pick<T>(List<WeightedChoice<T>> choices, Random rng)
        {
            if (choices == null || choices.Count == 0)
            {
                return default(T);
            }

            int totalWeight = choices.Sum(c => c.Weight);
            if (totalWeight <= 0)
            {
                return default(T);
            }

            int randomNumber = rng.Next(0, totalWeight);
            int cumulativeWeight = 0;

            foreach (var choice in choices)
            {
                cumulativeWeight += choice.Weight;
                if (randomNumber < cumulativeWeight)
                {
                    return choice.Item;
                }
            }

            return choices.Last().Item;
        }
    }
}
